XWIKI-15444: Fix export UI

surli opened the pull request:
## Scope of the PR

This PR aims at fixing issue XWIKI-15444, specifically 3 different issues have been listed:
 1. The export is not working when "More pages" pagination node is selected,
 2. The export is not working when lot of pages are selected,
 3. The export does not take into account the hidden pages

This PR targets all issues.

## Design

As for issues 1 and 2, the main change of the PR is to support in the Export API excluded pages, and to relate them to an included pages. It is then possible to build complex requests by properly ordering the argument: 

```
pages[0] = Foo.% // everything under Foo will be exported...
excludes[0] = Foo.Bar&Foo.Baz // ... except Bar and Baz pages

// the 3 following pages are included individually without excludes
pages[1] = Bar.WebHome
excludes[1] = 

pages[2] = Bar.Baz
excludes[2] = 

pages[3] = Bar.Foo
excludes[3] = 

pages[4] = Baz.% // everything under Baz is exported
excludes[4] = Baz.Bar.% // except all children of Baz.Bar including WebHome
```

Note that the excludes parameter is optional to respect backward compatibility of the API, however whenever it's used, it should always be accompanied by a pages argument, e.g.: 

```
// old API, still ok
pages[0] = Bar.%
pages[1] = Bar.Baz
pages[2] = Bar.Foo

// new API: OK
pages[0] = Bar.%
excludes[0] = Bar.Bar

pages[1] = Bar.Baz
excludes[1] = 

pages[2] = Bar.Foo
excludes[2] = 

// new API OK
pages[0] = Bar.%
excludes[0] = Bar.Bar

pages[1] = Bar.Baz
pages[2] = Bar.Foo

// new API NOT OK -> will throw an exception
pages[0] = Bar.%
excludes[0] = Bar.Bar
excludes[1] = Bar.Baz
``` 

Then the javascript associated to the tree modal as been rewritten to create proper hidden inputs on the fly to satisfy this new API given what has been selected or not (including choice made for pagination nodes).

Finally, regarding issue 3, the document tree is now displaying hidden files no matter what are the user preferences.

tmortagne commented on xwiki-platform-core/xwiki-platform-oldcore/pom.xml:
```diff
@@ -399,6 +399,21 @@
+    </dependency>
+    <dependency>
+      <groupId>org.xwiki.platform</groupId>
+      <artifactId>xwiki-platform-index-tree-api</artifactId>
```
You cannot add this dependency since it depends on oldcore. You will need to mock the implementation you need in it.

surli replied:
Even when it's a test dependency?

atallahade commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -268,8 +337,26 @@ private String exportXAR(XWikiContext context) throws XWikiException, IOExceptio
+        String uncheckedPagesRequestValue = request.get("unchecked-pages");
+        String otherPagesRequestValue = request.get("other-pages");
+
+        boolean all = ArrayUtils.isEmpty(checkedPages) && StringUtils.isEmpty(otherPagesRequestValue) && StringUtils.isEmpty(checkedPagesRequestValue);
```
This line is exceeding [120 characters](https://github.com/xwiki/xwiki-commons/blob/master/xwiki-commons-tools/xwiki-commons-tool-verification-resources/src/main/resources/checkstyle.xml#L214).

atallahade commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -268,8 +337,26 @@ private String exportXAR(XWikiContext context) throws XWikiException, IOExceptio
+        boolean all = ArrayUtils.isEmpty(checkedPages) && StringUtils.isEmpty(otherPagesRequestValue) && StringUtils.isEmpty(checkedPagesRequestValue);
+
+        boolean otherPages = false;
+        if (!StringUtils.isEmpty(otherPagesRequestValue)) {
```
Same comment as above.

atallahade commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -114,7 +119,26 @@ private String exportHTML(XWikiContext context) throws XWikiException, IOExcepti
+        String otherPagesRequestValue = request.get("other-pages");
+
+        boolean otherPages = false;
+        if (!StringUtils.isEmpty(otherPagesRequestValue)) {
```
From the javadoc `Boolean valueOf(String s)`:
> Returns a Boolean with a value represented by the specified string. The Boolean returned represents a true value if the string argument is not null and is equal, ignoring case, to the string "true".

So you might not need this line.

mflorea replied:
Yes, you can assign directly ``Boolean.valueOf()``.

mflorea commented on xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/uicomponents/exporter/exporter.js:
```diff
@@ -176,36 +180,90 @@ require([
+            if (treeReference.is_checked(node)) {
+                // if the node is pagination then we should set the otherPages attribute
+                if (node.data.type == "pagination") {
+                    otherPages.attr('value', 'true');
```
The recommended way to set the value is through `val()``.

mflorea commented on xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/uicomponents/exporter/exporter.js:
```diff
@@ -176,36 +180,90 @@ require([
+            } else {
+                // if the node is pagination then we should set the otherPages attribute
+                if (node.data.type == "pagination") {
+                    otherPages.attr('value', 'false');
```
Same as above. Any reason for not using ``val()``?

mflorea commented on xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/uicomponents/exporter/exporter.js:
```diff
@@ -176,36 +180,90 @@ require([
+                } else {
+                    var pageReference = XWiki.Model.resolve(node.data.id, XWiki.EntityType.DOCUMENT);
+                    checkedPages.push(XWiki.Model.serialize(pageReference));
+                    // In case the page coudl have children
```
Typo.

mflorea commented on xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/uicomponents/exporter/exporter.js:
```diff
@@ -176,36 +180,90 @@ require([
+                } else {
+                    var pageReference = XWiki.Model.resolve(node.data.id, XWiki.EntityType.DOCUMENT);
+                    uncheckedPages.push(XWiki.Model.serialize(pageReference));
+                    // In case the page coudl have children
```
Typo.

mflorea commented on xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/uicomponents/exporter/exporter.js:
```diff
@@ -176,36 +180,90 @@ require([
+                    uncheckedPages.push(XWiki.Model.serialize(pageReference));
+                    // In case the page coudl have children
+                    if (pageReference.getName() == 'WebHome') {
+                        // The node could be checked but not loaded
```
> // The node could be **un**checked but not loaded

mflorea commented on xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/uicomponents/exporter/exporter.js:
```diff
@@ -176,36 +180,90 @@ require([
             }
+          });
+
+          var aggregatePageNames = function (arrayOfNames) {
```
You can use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join instead.

mflorea commented on xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/uicomponents/exporter/exporter.js:
```diff
@@ -176,36 +180,90 @@ require([
+          };
+
+          checkedPagesInput.attr('value', aggregatePageNames(checkedPages));
+          uncheckedPagesInput.attr('value', aggregatePageNames(uncheckedPages));
```
Again, better use ``val()``. What happens if a page name contains ``&``. Any character is valid within the page name.

mflorea replied:
Yes, that should work.

mflorea commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -114,7 +119,26 @@ private String exportHTML(XWikiContext context) throws XWikiException, IOExcepti
+        }
+
+        if (!StringUtils.isEmpty(uncheckedPagesRequestValue)) {
+            uncheckedPages = uncheckedPagesRequestValue.split("&");
```
You make the assumption that page names can't contain ``&``, but they can.

mflorea commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -114,7 +119,26 @@ private String exportHTML(XWikiContext context) throws XWikiException, IOExcepti
+        }
+
+        if (!StringUtils.isEmpty(checkedPagesRequestValue)) {
+            checkedPages = checkedPagesRequestValue.split("&");
```
Same here, ``&`` is valid and not that uncommon in page names.

mflorea commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -179,7 +202,7 @@ private String exportHTML(XWikiContext context) throws XWikiException, IOExcepti
 
             DocumentReferenceResolver<String> resolver =
-                Utils.getComponent(DocumentReferenceResolver.TYPE_STRING, "current");
+                    Utils.getComponent(DocumentReferenceResolver.TYPE_STRING, "current");
```
I assume this is the result of the automatic formatter using the XWiki code style. It strange that it uses 8 spaces to indent.

mflorea commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -197,23 +220,69 @@ private String exportHTML(XWikiContext context) throws XWikiException, IOExcepti
     }
 
+    /**
+     * Resolve the set of pages to export based on choices made in a paginated tree
```
The XWiki code style leaves an empty line here. Did you configure your IDE to use the XWiki code style?

mflorea commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -197,23 +220,69 @@ private String exportHTML(XWikiContext context) throws XWikiException, IOExcepti
+            XWikiDocument doc = context.getDoc();
+
+            // use the document tree to get all the pages
+            Tree nestedpages = Utils.getComponent(Tree.class, "nestedPages");
```
Camel case makes the variable names easier to read.

mflorea commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -197,23 +220,69 @@ private String exportHTML(XWikiContext context) throws XWikiException, IOExcepti
+                formattedReferences.add(child.substring(index));
+            }
+
+            // we need to add the current document also
```
Why?

mflorea replied:
I think the reason is that currently you can't select "other pages" without selecting the parent page.

mflorea commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -197,23 +220,69 @@ private String exportHTML(XWikiContext context) throws XWikiException, IOExcepti
+            Tree nestedpages = Utils.getComponent(Tree.class, "nestedPages");
+            String nodeId = "document:" + doc.getDocumentReference().toString();
+            int childCount = nestedpages.getChildCount(nodeId);
+            List<String> children = nestedpages.getChildren(nodeId, 0, childCount);
```
This gives you the direct child nodes. How do you include all the descendant child nodes? (I mean all the sub-tree).

surli replied:
I suppose I don't right now: I did not check this scenario yet.
So 1. do you know if there's any way to get the descendant recursively from the tree API? 
   2. according to you, what is the minimum scenario to need an export with subtrees: defining a subwiki?

surli replied:
OK I'll check what happens with nested pages then.

surli replied:
Indeed there is a problem with nested pages, but I cannot process them as done in the javascript: in the JS we know which nodes are not loaded, so we can mark them with a joker to mark the subtree to be exported.
Here basically I need to process the nodes that are hidden: everyone of them might contain a subtree: so I would need to add the joker to everyone of them.

mflorea commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -268,8 +337,26 @@ private String exportXAR(XWikiContext context) throws XWikiException, IOExceptio
+
+        if (!StringUtils.isEmpty(checkedPagesRequestValue)) {
+            checkedPages = checkedPagesRequestValue.split("&");
+        }
```
The above lines are duplicated in the ``render()`` method.

surli commented:
Thanks @mflorea @atallahade for the review: I'll check my checkstyle config in the IDE, and try to add the corner-cases you mentioned in my test (page with `&` in the name, subtrees).

surli commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/export_modal.vm:
```diff
@@ -60,40 +60,46 @@
-          $services.rendering.render($exportExtension.execute(), 'html/5.0')
-        #end
+          <form method="post" id="export-form">
+              <input type="hidden" id="checked-pages" name="checked-pages" />
```
Just checked and I'm using the XWiki profile in the IDE but apparently it did not set the right indent for HTML, CSS, JS & Velocity.

tmortagne commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -114,7 +119,26 @@ private String exportHTML(XWikiContext context) throws XWikiException, IOExcepti
-        Collection<DocumentReference> pageList = resolvePagesToExport(request.getParameterValues("pages"), context);
+        String[] checkedPages = request.getParameterValues("pages");
+        String[] uncheckedPages = null;
+        String checkedPagesRequestValue = request.get("checked-pages");
```
If this data is supposed to be a list why invent your own syntax instead of using standard HTTP one and simply use getParameterValues like we do for pages ?

surli replied:
Because here the values are coming from a POST request: it would involve to create as many input field as I have pages checked/unchecked. 
Here it seems cleaner to me to only have 2 input fields and to parse their value: then I need to do it manually.

surli commented:
@mflorea this is ready for review again: I still have one test to add to check subtress. I already checked it manually so from my point of view the design won't change now.

tmortagne commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -94,6 +111,43 @@ public String render(XWikiContext context) throws XWikiException
+            String checkedPagesParam = request.get("checked-pages");
+
+            if (!StringUtils.isEmpty(checkedPagesParam)) {
+                this.checkedPages = checkedPagesParam.split("&");
```
As @mflorea pointed out in a previous comment this code seems to be impossible to have a page name containing `&` in that list. If you insist on inventing your own syntax you will need escaping support in this syntax because in XWiki there is no forbidden characters in names.

surli replied:
The names of pages are encoded in the exporter.js, see: https://github.com/xwiki/xwiki-platform/pull/855/files#diff-732756cd30b04d597a99718c625c573aR246

So the "&" which might exist in the filename are already encoded.
Now as I answered in your previous comment I cannot use the `getParamValues` as it would involve create as many input fields as I have checked pages: if you have any other advice to not use this syntax, I take it.

tmortagne replied:
The code would be much easier to understand if the decode was done here. So that checkedPages actually contain a list of pages :)

surli replied:
Indeed. I'll fix that.

surli commented:
@tmortagne as we discussed the design yesterday, could you review this one? Thanks.

surli commented:
@tmortagne I need to manage the hidden pages but you can already review it.

surli commented:
Hidden pages are now managed. Please read the description of the PR: it has been edited to explain the changes.

tmortagne commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -70,6 +79,10 @@
  */
 public class ExportAction extends XWikiAction
 {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExportAction.class);
```
Does not seems to be used.

tmortagne commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -70,6 +79,10 @@
 public class ExportAction extends XWikiAction
 {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExportAction.class);
+    private static final String TREE_DOCUMENT_PREFIX = "document:";
```
Same.

tmortagne commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -136,52 +225,100 @@ private String exportHTML(XWikiContext context) throws XWikiException, IOExcepti
+    private String extractWikiName(String pattern, XWikiContext context)
+    {
+        String wikiName;
+        if (pattern.contains(":")) {
```
You should do the if on `index` instead of parsing the String twice (only substring if index > 0).

tmortagne commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -195,11 +332,11 @@ private String exportHTML(XWikiContext context) throws XWikiException, IOExcepti
-                            "view", context.getUser(), pageReference, context))
-                        {
-                            pageList.add(resolver.resolve(pageReference));
+                        String pageReferenceStr = wikiName + XWikiDocument.DB_SPACE_SEP + docName;
```
Better pass a WikiReference as second parameter of the resolver (should always avoid directly manipulate reference syntax).

mflorea commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/export_modal.vm:
```diff
@@ -4,70 +4,82 @@
+      </div>
+      <div class="modal-body">
+        <div class="panel-group" id="exportModelAccordion" role="tablist" aria-multiselectable="true">
+          #set($exportExtensions = $services.uix.getExtensions('org.xwiki.plaftorm.menu.export.buttons'))
```
The HTML was already nicely indented. I don't understand why you removed the indentation.

surli replied:
Regarding all indentation problem: once I fixed my IDE I used the "format code" which completely mess up the diff. My bad.
I'll use the first version of the file and apply back my changes to fix that.

mflorea commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/export_modal.vm:
```diff
@@ -4,70 +4,82 @@
+      #end
+      #foreach ($exportExtension in $exportExtensions)
+        #if ($exportExtension.parameters.category == 'office')
+          $services.rendering.render($exportExtension.execute(), 'html/5.0')
```
Same here. Couldn't you fix the long lines without affecting the indentation of the other lines? It also makes the diff harder to read.

mflorea commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/export_modal.vm:
```diff
@@ -4,70 +4,82 @@
-      </h4>
-    </div>
-    <div id="exportModelOtherCollapse" class="panel-collapse collapse" role="tabpanel" aria-labelledby="exportModalHeadingOther">
-      <div class="panel-body">
```
Same. I think the previous indentation was good.

surli replied:
@mflorea here I just added the form around the div but apparently git has some trouble to process a good diff there...

mflorea replied:
You could try

    <form id="export-form" method="post" class="panel-body">

i.e. convert the panel body DIV into a form instead of wrapping it. This way you don't need to change the indentation.

surli replied:
Hm. Didn't thought about that. I'll try.

surli replied:
Indeed it fixes the diff.

mflorea commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/export_modal.vm:
```diff
@@ -4,70 +4,82 @@
-              <p class="text-right"><a href="#" class="tree_select_all">$escapetool.xml($services.localization.render('core.exporter.selectAll'))</a> / <a href="#" class="tree_select_none">$escapetool.xml($services.localization.render('core.exporter.selectNone'))</a></p>
+              <p class="text-right"><a href="#" class="tree_select_all">$escapetool.xml($services.localization.render(
+                'core.exporter.selectAll'))</a> / <a href="#"
+                                                     class="tree_select_none">$escapetool.xml($services.localization.render(
```
Strange indentation.

mflorea commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/export_modal.vm:
```diff
@@ -78,23 +90,23 @@
           </div>
         #end
-        <a href="$doc.getURL('export', "format=html&amp;$!escapetool.html($request.getQueryString())")" class="btn btn-primary" rel="nofollow">$services.localization.render('core.menu.export.html')</a>
+        <a href="$doc.getURL('export', "format=html")" class="btn btn-primary"
```
You dropped the query string from the current URL. Where did it go? You don't need it at all?

surli replied:
No it's now completely computed through the JS. But we might keep it, as the URL is rewritten in the JS, so I can push it back.

mflorea commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/export_modal.vm:
```diff
@@ -78,23 +90,23 @@
-            #set($exportURL = "$exportURL&amp;pages=${escapetool.url($services.model.serialize($doc.documentReference.parent,'default'))}.WebPreferences")
-          #end
-          <a href="$doc.getURL('export', "$exportURL&amp;$!docvariant")" class="btn btn-primary" rel="nofollow">$services.localization.render('core.menu.export.xar')</a>
+          #set($exportURL = "format=xar&amp;name=$escapetool.url($doc.fullName)")
```
You dropped the ``$!docvariant``. Was it not needed?

surli replied:
That's something I actually did not understood in the file: did not find any usage or declaration of this neither in the .vm or in the ExportAction API. If you have any clue I take it.

mflorea replied:
* ``view.vm`` includes ``startpage.vm`` includes ``xwikivars.vm``
* ``view.vm`` includes ``contentview.vm`` includes ``contentheader.vm`` includes ``menus_content.vm`` includes ``export_modal.vm``

``$docvariant`` defined in ``xwikivars.vm`` is available in ``export_modal.vm``.

surli replied:
Thanks! But then it's strange because it seems that XAR export in ExportAction does not handle revision or language and from what I see in `xwikivars` docvariant is used only for those two parameters.

surli replied:
So should we keep it?

surli commented:
For me this one is finished @tmortagne @mflorea

tmortagne commented:
Fine with me (but did not really looked at client side).

atallahade commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-test/xwiki-platform-flamingo-skin-test-tests/src/test/it/org/xwiki/flamingo/test/ui/XARExportTest.java:
```diff
@@ -0,0 +1,249 @@
+        // Create a space Foo
+        getUtil().createPage("Foo", "WebHome", "Foo", "Foo");
+
+        // Create 100 pages under that space (will be used for the test with lot of selected pages)
```
Do you need that many pages? Each page being saved takes some time AFAIK, unless you can do it in batch or in a database session.

surli replied:
I'm using the REST API here so it does not take so much time (the test were taking around 20 seconds, once launched).
I need lot of pages to check the behaviour with a long list of pages selected, and I also want to check the "more pages" node, but indeed 100 is quite a lot. I'll create 50.

atallahade replied:
It seems that you haven't updated the comment (and it's great that the pages creation is fast enough!).

atallahade commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-test/xwiki-platform-flamingo-skin-test-tests/src/test/it/org/xwiki/flamingo/test/ui/XARExportTest.java:
```diff
@@ -0,0 +1,249 @@
+        viewPage.clickMoreActionsSubMenuEntry("tmExport");
+
+        ExportModal exportModal = new ExportModal();
+        //assertTrue(exportModal.isDisplayed());
```
Leftover.

atallahade commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -70,6 +75,117 @@
+        /**
+         * All patterns that have to be included in the request
+         *
+         * Warning: order of the arguments matters, it's linked with excludedPages
```
In that case you can maybe use [`org.apache.commons.lang3.tuple.Pair`](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/tuple/Pair.html) or a HashMap?

atallahade commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -70,6 +75,117 @@
+
+            if (!this.excludedPages.isEmpty() && this.includedPages.size() < this.excludedPages.size()) {
+                throw new XWikiException(XWikiException.MODULE_XWIKI_APP, XWikiException.ERROR_XWIKI_APP_EXPORT,
+                    "You cannot more excludes argument than pages argument. See the export documentation.");
```
Typo:
> You cannot **have** more excludes arguments than pages arguments

atallahade commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -94,6 +210,8 @@ public String render(XWikiContext context) throws XWikiException
     }
 
+
+
```
Code style: you should remove this extra line.

atallahade commented on xwiki-platform-core/xwiki-platform-oldcore/src/test/java/com/xpn/xwiki/web/ExportActionTest.java:
```diff
@@ -122,12 +142,109 @@ public void exportFullSpaceUsingWildcardsAsXAR() throws Exception
+        when(request.getCharacterEncoding()).thenReturn("UTF-8");
+        // Export all pages in the "Space" space
+        when(request.getParameterValues("pages")).thenReturn(new String[] {"Space.%25", "Foo.%25", "Bar.Baz.%25"});
+        when(request.getParameterValues("excludes")).thenReturn(new String[] { "Space.Page1&Space.Page2", "", "Bar.Baz.WebHome" });
```
Code style: line is too long.

surli commented:
@atallahade 
> The way excludes and pages arguments are handle is a bit complicated IMO. Wouldn't it be simpler to "untie" the two arguments (and give more freedom to the way they can be used), i.e. having a list of "excludes" independent of the "pages" list.

Looking for a counterexample but I cannot find any right now, so I guess it should work. WDYT @tmortagne?

tmortagne commented:
@surli as we discussed applying all the excludes to all the pages is a problem when the excludes are patterns. What @atallahade propose (which is what you did yourself in a previous version) only work if you forbid any pattern other than something ending with %.

atallahade commented:
@tmortagne Could you give an example? I'm not sure I understand why having an exclude pattern ending with `%` might cause problems.

tmortagne commented:
>  I'm not sure I understand why having an exclude pattern ending with % might cause problems

That's because I said the opposite :)

Imagine that you export two spaces but in one of them you want to exclude `%.Toto` (but only in that space). It's not a problem for patterns where the % (if any) is only at the end.

atallahade commented:
@tmortagne 
> Imagine that you export two spaces but in one of them you want to exclude %.Toto (but only in that space). It's not a problem for patterns where the % (if any) is only at the end.

If you want to exclude `%.Toto` in a space, why not write `Space.%.Toto`?

tmortagne commented:
> If you want to exclude %.Toto in a space, why not write Space.%.Toto?

This is not enough, you also need to exclude `Space.Toto`.

You are the one talking about simplicity but I find much simpler to tie excludes to pages because that's actually what you want to do 90% of the time (exclude stuff from a specific `pages` element) and you don't have to wonder how to craft something which is not going to mess with other `pages` elements that don't you care about.

atallahade commented:
> you don't have to wonder how to craft something which is not going to mess with other pages elements that don't you care about.

Good point, it might actually be a good trade off to restrict the excludes pattern to spaces over than having more freedom but also having to deal with tricky cases.

surli commented:
@mflorea could you review the JS of this PR?

mflorea commented on xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/uicomponents/exporter/exporter.js:
```diff
@@ -152,60 +154,188 @@ require([
-          var checkedNodes = treeReference.get_checked(true);
-          if (url.params === undefined) {
-            url.params = {'pages': []};
+        var processLevel = function(rootNode, arrayOfPages, arrayOfExcludes) {
```
If I understood correctly, the items from ``arrayOfExcludes`` are mapped to the items from ``arrayOfPages`` so it would have been nice to use a map data type to make this more explicit. Unfortunately [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) is not supported on older version of Internet Explorer and the plain JavaScript object has some reserved keys that theoretically could collide with the page references. In practice the page reference should contain at least one dot so there should be no collision. The only issues could be the page order, but I don't think it matters (Map preserves the key insertion order but we can't use it yet..)

surli replied:
OK I'll refactor to use a POJO then.

mflorea commented on xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/uicomponents/exporter/exporter.js:
```diff
@@ -152,60 +154,188 @@ require([
+          // if we don't have a pagination node, or we have one and it's checked:
+          // we manage the export with exporting the space and excluding stuff that needs to be
+          if (!isPaginationExisting || (isPaginationExisting && isPaginationChecked)) {
+            includedPages = [ pageJoker ];
```
What's the point of setting ``includedPages`` here? You don't use it after this.

surli replied:
indeed. Old stuff I guess. Will remove it.

surli commented on xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/uicomponents/exporter/exporter.js:
```diff
@@ -152,60 +154,188 @@ require([
+         * FTR the parameters are ordered as the first element of pages is linked to the first element of excludes.
+         *
+         * @param rootNode: the parent node to process
+         * @param arrayOfPages: a list of list of pages to includes
```
Indeed it changes after a second round. I'll fix the doc.

surli commented:
OK I changed both the java and javascript side to use a Map instead of two lists for included/excluded pages, as suggested by @atallahade and @mflorea. 
So for me all suggestions have been taken into account and fixed.

surli commented:
> We'll need to test it properly.

I added some functional test to check the behaviour as well as a new unit test. Now, I also tested it manually. 
Do you have another kind of test in mind?

mflorea commented:
> Do you have another kind of test in mind?

The QA team, but after we merge it :)

tmortagne commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ExportAction.java:
```diff
@@ -70,6 +75,93 @@
+            String[] pages = request.getParameterValues("pages");
+            String[] excludes = request.getParameterValues("excludes");
+
+            this.exportPages = new HashMap<>();
```
Just a detail: I would use a LinkedHashMap here to keep the same order than what you got in input.

surli replied:
Fixed.
